internal/Yesod/Session/Memcache/Storage.hs

```diff
@@ -0,0 +1,91 @@
+module Yesod.Session.Memcache.Storage
+  ( memcacheStorage
+  , SessionPersistence (..)
+  ) where
+
+import Internal.Prelude
+
+import Control.Monad.Reader (MonadReader (ask))
+import Database.Memcache.Client qualified as Memcache
+import Database.Memcache.Types qualified as Memcache
+import Session.Key
+import Yesod.Session.SessionType
+import Yesod.Session.Storage.Exceptions
+import Yesod.Session.Storage.Operation
+
+-- | Mapping between 'Session' and Memcache representation.
+data SessionPersistence = SessionPersistence
+  { databaseKey :: SessionKey -> Memcache.Key
+  , toDatabase :: Session -> Memcache.Value
+  , fromDatabase :: Memcache.Value -> Session
+  , runDB :: forall a b m. ReaderT b IO a -> m a
+  }
+
+memcacheStorage
+  :: forall m result
+   . (MonadReader Memcache.Client m, MonadIO m)
+  => SessionPersistence
+  -> StorageOperation result
+  -> m result
+memcacheStorage sp@SessionPersistence {} = \case
+  GetSession sessionKey ->
+    let get client = Memcache.get client (sp.databaseKey sessionKey)
+    in  (sp.fromDatabase . fstOf3) <$$> (ask >>= liftIO . get)
```

Comment:
I think I'd rather see a little noise here than need to add custom prelude utilities to this little project.

```haskell
fmap (sp.fromDatabase . (\(x, _, _) -> x)) <$> (ask >>= liftIO . get)
```

Also though I think it's easier to follow if laid out a little more verbosely in a `do` block?

```haskell
  GetSession sessionKey -> do
    client <- ask
    result <- liftIO $ Memcache.get client $ sp.databaseKey sessionKey
    pure $ result <&> \(value, _flags, _version) -> sp.fromDatabase value
```